PathSegment[] pastSegments
PathSegment[] futureSegments
PathSegment[] filteredSegments
PathSegment[] bestSolution